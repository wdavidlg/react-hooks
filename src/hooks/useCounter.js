import { useState } from "react"

export const useCounter = (initialState = 10) => {

    const [counter, setCounter] = useState(initialState)

    const increment = () => {
        setCounter( prev => prev + 1)
    }

    const decrement = () => {
        setCounter(prev => prev - 1)
    }

    const reset = () => {
        setCounter(initialState);
    }

    return {
        counter,
        increment,
        decrement,
        reset
    };
}
