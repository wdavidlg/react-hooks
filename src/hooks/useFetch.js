import { useEffect, useRef, useState } from "react";


export const useFetch = (id) => {

    const isMounted = useRef(true);

    const base = 'https://breakingbadapi.com/api/';
    const endpoint = 'quotes/';
    const url = `${base}${endpoint}${id}`;

    const [state, setState] = useState({
        data: null, 
        loading: true,
        error: null
    })

    useEffect(() => {
        console.log('Mounted...');
        return () => {
            isMounted.current = false;
            console.log('Paso a falso', isMounted)
        }
    }, [])

    useEffect(() => {

        setState({
            data: null, 
            loading: true,
            error: null
        })

        fetch(url)
            .then( resp => resp.json())
            .then( data => {
                setTimeout(() => {

                    if(isMounted.current){
                        console.log('aaaaa-',isMounted)
                        try{
                            if(data.name){
                                throw 'No hay info';
                            }
                            setState({
                                loading: false,
                                error: null,
                                data
                            })
                        }catch(err){
                            setState({
                                loading: false,
                                error: err,
                                data: null
                            })
                        }
                    }else{
                        console.log('setState no se llam√≥')
                    }
                }, 1000)
            })
    }, [url])

    return state;

}
