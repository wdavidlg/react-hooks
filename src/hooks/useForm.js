import { useState } from "react"


export const useForm = (initialState = {}) => {

    const [values, setValues] = useState(initialState);

    const handleInputChange = (e) => {
        const {value} = e.target;
        const {name} = e.target;
        setValues((prev) => ({...prev, [name]: value}))
    }

    const reset = () => {
        setValues(initialState)
    }

    return [values, handleInputChange, reset];
}
