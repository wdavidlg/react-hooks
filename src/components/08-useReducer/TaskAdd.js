import React from 'react'
import { useForm } from "../../hooks/useForm";

const TaskAdd = ({handleAdd}) => {
    const [{ task }, handleInputChange, reset] = useForm({
        task: "",
      });

      const handleSubmit = (e) => {
        e.preventDefault();
        if (task.trim().length === 0) return;
        const newTask = {
          id: new Date().getTime(),
          task: task,
          done: false,
        };
        handleAdd(newTask);
        reset();
      };

    return (
        <>
          <h4>Add Task</h4>
          <hr />
          <form onSubmit={handleSubmit}>
            <input
              onChange={handleInputChange}
              value={task}
              type="text"
              name="task"
              className="form-control"
              placeholder="Task"
              autoComplete="off"
            />
            <button type="submit" className="btn btn-outline-primary btn-add">
              Add
            </button>
          </form>  
        </>
    )
}

export default TaskAdd
