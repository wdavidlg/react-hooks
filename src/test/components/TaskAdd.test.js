import {shallow} from 'enzyme'
import TaskAdd from '../../components/08-useReducer/TaskAdd'
import { useForm } from '../../hooks/useForm';
//jest.mock('../../hooks/useForm')

describe('Pruebas en <TodoAdd />', () => {

    const handleAdd = jest.fn();
    // let wrapper = null;
    
    // beforeEach(() => {
    //     jest.clearAllMocks();
    //     useForm.mockReturnValue([{
    //         task: ''
    //     }, () => {}, () => {}]);

    //     wrapper = shallow(
    //         <TaskAdd 
    //             handleAdd={handleAdd}
    //         />
    //     )
    // })

    // test('debe de mostrarse correctamente', () => {

    //     expect(wrapper).toMatchSnapshot();
        
    // })

    // test('No debe de llamar a handleAdd', () => {
        
    //     wrapper.find('form').simulate('submit', {
    //         preventDefault: ()=>{}
    //     })
    //     expect(handleAdd).not.toHaveBeenCalled();
        
    // })
    
    // test('debe de llamar a handleAdd', () => {

    //     useForm.mockReturnValue([{
    //         task: 'Aprender React'
    //     }, () => {}, () => {}]);
        
    //     const wrapper = shallow(<TaskAdd  handleAdd={handleAdd}/>)
    //     wrapper.find('form').simulate('submit', {
    //         preventDefault: ()=>{}
    //     })
    //     expect(handleAdd).toHaveBeenCalledTimes(1);
    //     expect(handleAdd).toHaveBeenCalledWith({
    //         done: false,
    //         task: 'Aprender React',
    //         id: expect.any(Number)
    //     })

    // })

    test('Otra forma de hacerlo todo, yver si el reset funciona', () => {
        
        const value = 'Aprender Node';
        
        const wrapper = shallow(<TaskAdd  handleAdd={handleAdd}/>)
        //Debe de mostrarse correctamente
        expect(wrapper).toMatchSnapshot();
        
        //No debe de llamar a handleAdd
        wrapper.find('form').simulate('submit', {preventDefault(){}})
        expect(handleAdd).not.toHaveBeenCalled();

        //Debe de llamar a handleAdd
        wrapper.find('input').simulate('change', {
            target:{
                value,
                name: 'task'
            }
        })
        wrapper.find('form').simulate('submit', {preventDefault(){}})
        expect(handleAdd).toHaveBeenCalledTimes(1);
        //Debe de resetear el valor
        expect(wrapper.find('input').prop('value')).toBe('')

    })
    
    
    
})
