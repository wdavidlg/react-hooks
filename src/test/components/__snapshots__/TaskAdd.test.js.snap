// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAdd /> Otra forma de hacerlo todo, yver si el reset funciona 1`] = `
<Fragment>
  <h4>
    Add Task
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="task"
      onChange={[Function]}
      placeholder="Task"
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary btn-add"
      type="submit"
    >
      Add
    </button>
  </form>
</Fragment>
`;
